### PR TITLE
🚨🚨🚨 Replace DataLoader logic for Accelerate in Trainer, remove unneeded tests 🚨🚨🚨

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR:
- Guts the internals for the `DataLoader` in all basic distributed fashions (replacing `pl.Loader` for TPU coming in a follow-up PR) to replace it with `accelerator.prepare`
- Removes **two** tests that were deemed unnecessary
  - Test 1 removed: `tests/trainer/test_trainer.py::TrainerIntegrationTest::test_sampler_seed`, deemed to no longer be necessary to reset the seed, as Accelerate's dataloader setup doesn't need any extra help when iterating/loading back in the seed, regardless of the torch version
  - Test 2 removed: `tests/trainer/test_trainer.py::TrainerIntegrationTest::test_training_finite_iterable_dataset`, as with Accelerate's new sampler for `IterableDataset` we'll actually catch if it's `None` and raise an error, a new test will be made + clear error message on the `Accelerate` side, with a test added to `Trainer` afterwards. 
- Modifies two tests to use the proper attribute: Accelerator's `DataLoaders` all have `total_batch_size` rather than `batch_size`
  - `test_train_and_eval_dataloaders` and `test_data_is_not_parallelized_when_model_is_parallel`

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger @pacman100

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1849,26 +1742,11 @@ def _inner_training_loop(
...
-                    # We just need to begin an iteration to create the randomization of the sampler.
-                    # That was before PyTorch 1.11 however...
-                    for _ in train_dataloader:
-                        break
-                else:
-                    # Otherwise we need to call the whooooole sampler cause there is some random operation added
-                    # AT THE VERY END!
-                    _ = list(train_dataloader.sampler)
+                for _ in train_dataloader:
+                    break
```
Comment (sgugger) [inline src/transformers/trainer.py:1746]: Is this still necessary with the RNG reloading?
Edit: ah looks like we are not doing that yet.

**Review by sgugger — APPROVED** [review]
Thanks a ton, LGTM!

File: src/transformers/trainer.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/trainer.py:1746]: Yep not yet, so for now it's needed :)

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3205,72 +3083,42 @@ def evaluation_loop(
 
             # Update containers on host
             if loss is not None:
-                losses = self._nested_gather(loss.repeat(batch_size))
-                losses_host = losses if losses_host is None else torch.cat((losses_host, losses), dim=0)
+                losses = self.accelerator.gather_for_metrics((loss.repeat(batch_size)))
```
Comment (pacman100) [inline src/transformers/trainer.py:3086]: Nice!

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3205,72 +3083,42 @@ def evaluation_loop(
...
-                logits = self._nested_gather(logits)
+                logits = self.accelerator.gather_for_metrics((logits))
                 preds_host = logits if preds_host is None else nested_concat(preds_host, logits, padding_index=-100)
-            if labels is not None:
-                labels = self._nested_gather(labels)
-                labels_host = labels if labels_host is None else nested_concat(labels_host, labels, padding_index=-100)
-            self.control = self.callback_handler.on_prediction_step(args, self.state, self.control)
-
-            # Gather all tensors and put them back on the CPU if we have done enough accumulation steps.
-            if args.eval_accumulation_steps is not None and (step + 1) % args.eval_accumulation_steps == 0:
```
Comment (pacman100) [inline src/transformers/trainer.py:3232]: `args.eval_accumulation_steps` is currently being ignored

**Review by pacman100 — COMMENTED** [review]
Thank you @muellerzr for all the cool work on using Accelerate for preparing dataloaders, using `accelerator.gather_for_metrics` and `accelerator.pad_across_processes` for further simplification. 🚀

Left a couple comments regarding few training arguments being ignored.

File: src/transformers/trainer.py (hunk 3 above)
Comment (muellerzr) [reply src/transformers/trainer.py:3232]: Yes, it's no longer needed here (gradient accumulation tests all pass). There's no need for the repeat logic with how the new gather_for_metrics/padding operates, unless you see otherwise

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -805,20 +800,6 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
...
-            # for backwards compatibility, we generate a seed here (which is sampled from a generator seeded with
-            # `args.seed`) if data_seed isn't provided.
-            # Further on in this method, we default to `args.seed` instead.
-            if self.args.data_seed is None:
-                seed = int(torch.empty((), dtype=torch.int64).random_().item())
-            else:
-                seed = self.args.data_seed
-            generator.manual_seed(seed)
-
-        seed = self.args.data_seed if self.args.data_seed is not None else self.args.seed
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/transformers/trainer.py:820]: Nope, Accelerate handles that better and doing this actually hinders reproducibility in that sense. (We use an entirely different dataloader/sampler mechanic than the one used here, so our version of seed isn't the same as what's being used for the samplers/dataloaders here)

File: src/transformers/trainer.py (hunk 3 above)
Comment (muellerzr) [reply src/transformers/trainer.py:3232]: Yes, this is needed else the metrics stay on the GPU using more ram during accum steps :)

File: src/transformers/trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/transformers/trainer.py:820]: The *tests* haven't shown it to, because (from what Sylvain and I were able to find), our dataloaders don't play with the same random seed setting that these do/applying generators. Keeping this in as a generator also causes tests to fail, showing that it's detrimental in that sense now with our new DataLoader setup. 

Also it's used when `__iter__` through the dataset here: https://github.com/huggingface/transformers/blob/main/src/transformers/trainer_pt_utils.py#L665-L667. Because we don't use a torch `DistributedSampler`, I don't believe it applies to us here. 

If later down the road someone points to a reproducible issue to this, I'll review it again but in my deep dive last week I didn't find an issue.

**Review by pacman100 — APPROVED** [review]
Thank you for your responses to the comments and for iterating on the PR. Super cool!

**Comment by franz101** [comment]
Great PR, currently this is breaking my custom collate_fn in the dataloader, still trying to understand why that is. First assumption might be due to multiprocessing?

**Comment by muellerzr** [comment]
@franz101 please open an issue with a reproducer of what you are trying to do so we can help :)
